<a href="https://colab.research.google.com/github/edsrisk/portfolio_optimization/blob/main/portfolio_opt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import libraries and dependencies
from pandas_datareader import data as web
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')
from pathlib import Path
from google.colab import drive
drive.mount('/content/drive')

In [7]:
from google.colab import files
files.upload()


Saving portfolio_data.csv to portfolio_data.csv


{'portfolio_data.csv': b'Date,BP,GS,HD,HP,MMM,NEU,UNH\r\n4/1/2021,24.290878,326.537201,306.144653,28.190351,191.292313,380.847137,365.730499\r\n4/5/2021,23.974642,322.450989,313.75473,26.614309,193.545746,386.46524,368.829132\r\n4/6/2021,24.804762,325.959137,312.143188,26.653957,193.416687,386.355835,363.040344\r\n4/7/2021,25.071587,325.450836,310.840027,27.28834,193.525879,382.060272,364.305725\r\n4/8/2021,24.42923,330.025421,312.789795,26.663868,194.290268,380.966461,363.558472\r\n4/9/2021,24.103111,329.696503,317.564758,25.593349,196.553604,384.058899,374.926788\r\n4/12/2021,24.330408,330.723053,319.116638,25.702385,196.384842,383.61145,374.906891\r\n4/13/2021,24.330408,326.577057,318.728668,25.841154,195.03479,380.330078,373.133392\r\n4/14/2021,25.239586,334.221252,318.350616,26.862114,195.312729,382.636963,374.259277\r\n4/15/2021,24.903585,337.410461,321.056458,26.544924,196.037399,382.139801,388.586792\r\n4/16/2021,24.883821,341.157806,326.368591,26.406151,197.129364,385.500702,3

In [8]:
# Access datafile
portfolio_df = pd.read_csv('portfolio_data.csv', index_col=0)
portfolio_df.head()

BP          GS          HD  ...         MMM         NEU         UNH
Date                                         ...                                    
4/1/2021  24.290878  326.537201  306.144653  ...  191.292313  380.847137  365.730499
4/5/2021  23.974642  322.450989  313.754730  ...  193.545746  386.465240  368.829132
4/6/2021  24.804762  325.959137  312.143188  ...  193.416687  386.355835  363.040344
4/7/2021  25.071587  325.450836  310.840027  ...  193.525879  382.060272  364.305725
4/8/2021  24.429230  330.025421  312.789795  ...  194.290268  380.966461  363.558472

[5 rows x 7 columns]

In [11]:
# Assign weights to stocks
weights = np.array([0.142587143, 0.142587143, 0.142587143, 0.142587143, 0.142587143, 0.142587143, 0.142587143])
weights

array([0.14258714, 0.14258714, 0.14258714, 0.14258714, 0.14258714,
       0.14258714, 0.14258714])

In [12]:
# checkpoint: make sure all weights add up to 1
sum = 0
for stock in weights:
    sum += stock
print(sum)

0.9981100009999999


In [13]:
# Calculate daily returns
stock_returns = portfolio_df.pct_change()

In [14]:
stock_returns.head()

BP        GS        HD        HP       MMM       NEU       UNH
Date                                                                          
4/1/2021       NaN       NaN       NaN       NaN       NaN       NaN       NaN
4/5/2021 -0.013019 -0.012514  0.024858 -0.055907  0.011780  0.014752  0.008472
4/6/2021  0.034625  0.010880 -0.005136  0.001490 -0.000667 -0.000283 -0.015695
4/7/2021  0.010757 -0.001559 -0.004175  0.023801  0.000565 -0.011118  0.003486
4/8/2021 -0.025621  0.014056  0.006273 -0.022884  0.003950 -0.002863 -0.002051

In [15]:
# Calculate the annualized covariance matrix
port_cov = stock_returns.cov() * 252
port_cov

BP        GS        HD        HP       MMM       NEU       UNH
BP   0.080086  0.038120 -0.000710  0.105238  0.008790  0.023773  0.003054
GS   0.038120  0.054952  0.014387  0.066179  0.015933  0.023774  0.008257
HD  -0.000710  0.014387  0.033892 -0.006335  0.018701  0.019425  0.017980
HP   0.105238  0.066179 -0.006335  0.273184  0.016043  0.026323  0.002059
MMM  0.008790  0.015933  0.018701  0.016043  0.025005  0.017243  0.015668
NEU  0.023773  0.023774  0.019425  0.026323  0.017243  0.065940  0.015819
UNH  0.003054  0.008257  0.017980  0.002059  0.015668  0.015819  0.029591

In [16]:
# Calculate portfolio variance
port_var = np.dot(weights.T, np.dot(port_cov, weights))
port_var

0.0297260666628749

In [17]:
# Calculate Volatility
port_vol = np.sqrt(port_var)
port_vol

0.17241248986913593

In [19]:
# Calculate annual portfolio return ) Weighted average returns
portfolio_ret = np.sum(stock_returns.mean() * weights) * 252
portfolio_ret

0.23395927260614777

In [20]:
print(" Portfolio Summary")
print("----------------------------")
#print("Assets: ", portfolio_df[:0])
print('Annual Return: ', round(portfolio_ret, 2) * 100, '% / Volatility: ', round(port_vol, 2) * 100, '% / Variance: ', round(port_var, 2) * 100, '%')

 Portfolio Summary
----------------------------
Annual Return:  23.0 % / Volatility:  17.0 % / Variance:  3.0 %


In [24]:
!pip install PyPortfolioOpt

     |████████████████████████████████| 60 kB 3.3 MB/s 
     |████████████████████████████████| 2.7 MB 14.2 MB/s 
  Attempting uninstall: cvxpy
    Found existing installation: cvxpy 1.0.31
    Uninstalling cvxpy-1.0.31:
      Successfully uninstalled cvxpy-1.0.31


In [25]:
from pypfopt.efficient_frontier import EfficientFrontier
from pypfopt import risk_models
from pypfopt import expected_returns

Portfolio Analysis: Optimization

In [26]:
mu = expected_returns.mean_historical_return(portfolio_df) # Expected return
S = risk_models.sample_cov(portfolio_df) # Annaulized covariance matrix

In [27]:
# Efficient Frontier: Max Sharpe Ratio - amount of return for a given amount of volatility

ef = EfficientFrontier(mu, S)
weights = ef.max_sharpe()
cleaned_weights = ef.clean_weights()
print(cleaned_weights)
ef.portfolio_performance(verbose=True)

OrderedDict([('BP', 0.0), ('GS', 0.5183), ('HD', 0.0), ('HP', 0.00238), ('MMM', 0.0), ('NEU', 0.0), ('UNH', 0.47931)])
Expected annual return: 57.9%
Annual volatility: 16.1%
Sharpe Ratio: 3.48


(0.5789041849458428, 0.16072604706714524, 3.4773715595229806)

In [28]:
import cvxpy as cp

In [29]:
# Rebalance portfolio, asset allocation

from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices
latest_prices = get_latest_prices(portfolio_df)
weights = cleaned_weights # rebalanced/optimized weights
discrete_allocation = DiscreteAllocation(weights, latest_prices, total_portfolio_value = 50000)
allocation, leftover = discrete_allocation.lp_portfolio()
print(allocation)
print('Funds Remaining: ${:.2f}'.format(leftover))

{'GS': 70, 'UNH': 60}
Funds Remaining: $30.60


Optimal portfolio will have 70 shares of Goldman Sachs and 60 shares of United Healthcare; The efficient frontier shows a return of 57.9% and volatility of 16.1% with sharpe ratio of 3.48